### ISSUE TITLE
Killed a build, now Spack thinks it is installed

### ISSUE THREAD

**Issue by adamjstewart** [LLNL/spack#2634]
I just tried building a package, noticed something was wrong during configure, and killed it with <kbd>Ctrl</kbd>+<kbd>C</kbd>. When I ran `spack purge` and tried installing it again, I saw:
```
==> parallel-netcdf is already installed in /blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/intel-17.0.0/parallel-netcdf-1.6.1-fq2r43zk2wcbpl3hrc5h6sfz2dwajbk7
Traceback (most recent call last):
  File "/soft/spack-0.10.0/bin/spack", line 198, in <module>
    main()
  File "/soft/spack-0.10.0/bin/spack", line 175, in main
    return_val = command(parser, args)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/cmd/install.py", line 346, in install
    package.do_install(**kwargs)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/package.py", line 1161, in do_install
    rec = spack.store.db.get_record(self.spec)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/database.py", line 81, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/database.py", line 554, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/database.py", line 549, in _get_matching_spec_key
    raise KeyError("No such spec in database! %s" % spec)
KeyError: 'No such spec in database! parallel-netcdf@1.6.1%intel@17.0.0 cflags="-axCORE-AVX2,AVX" cxxflags="-axCORE-AVX2,AVX" fflags="-axCORE-AVX2,AVX"+cxx+fortran+fpic arch=linux-centos6-x86_64^bison@3.0.4%intel@17.0.0 cflags="-axCORE-AVX2,AVX" cxxflags="-axCORE-AVX2,AVX" fflags="-axCORE-AVX2,AVX" arch=linux-centos6-x86_64^libpciaccess@0.13.4%intel@17.0.0 cflags="-axCORE-AVX2,AVX" cxxflags="-axCORE-AVX2,AVX" fflags="-axCORE-AVX2,AVX" arch=linux-centos6-x86_64^libsigsegv@2.10%gcc@6.1.0 arch=linux-centos6-x86_64^libtool@2.4.6%gcc@6.1.0 arch=linux-centos6-x86_64^m4@1.4.17%gcc@6.1.0+sigsegv arch=linux-centos6-x86_64^mvapich2@2.2%intel@17.0.0 cflags="-axCORE-AVX2,AVX" cxxflags="-axCORE-AVX2,AVX" fflags="-axCORE-AVX2,AVX"~debug~gforker+hydra+mrail~nemesis~nemesisib~nemesisibtcp~psm~remshell~slurm~sock arch=linux-centos6-x86_64^pkg-config@0.29.1%gcc@6.1.0+internal_glib arch=linux-centos6-x86_64^util-macros@1.19.0%gcc@6.1.0 arch=linux-centos6-x86_64'
```
When I ran `spack reindex` and tried installing again, I saw:
```
==> parallel-netcdf is already installed in /blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/intel-17.0.0/parallel-netcdf-1.6.1-fq2r43zk2wcbpl3hrc5h6sfz2dwajbk7
```
Of course, that directory is empty except for `.spack/spec.yaml`.

**Comment by tgamblin** [comment]
I don't see how this would happen unless the `ctrl-C` happened right between the setting of `keep_prefix` and `spack.store.db.add` in `Package.do_install()`.  We should probably turn off interrupts during that part of install.

**Comment by adamjstewart** [comment]
It didn't. It was mid-configure.

**Comment by adamjstewart** [comment]
Nope, I didn't. There were no error messages at all either, which was weird. Usually I think it prints a Spack error message saying that the user killed the command.

**Comment by jrood-nrel** [comment]
I experience this almost every time I `qdel` a batch submitted build script and have to delete an empty install prefix directory.

**Comment by tgamblin** [comment]
@jrood-nrel: I believe that `qdel` will kill the job and not let it clean up.  Spack relies on being able to execute the `finally` block... do you know what signal `qdel` sends?

**Comment by jrood-nrel** [comment]
@tgamblin Looks like:

> A  batch  job  being deleted by a server will be sent a SIGTERM signal following by a SIGKILL signal.

**Comment by adamjstewart** [comment]
This happened to me again today.
```
$ spack install -v --run-tests mpich %intel
==> mpich is already installed in /blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/intel-17.0.0/mpich-3.2-tg45j2rspl4sb35o32beqrwffrkyuopi
Traceback (most recent call last):
  File "/soft/spack-0.10.0/bin/spack", line 198, in <module>
    main()
  File "/soft/spack-0.10.0/bin/spack", line 175, in main
    return_val = command(parser, args)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/cmd/install.py", line 346, in install
    package.do_install(**kwargs)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/package.py", line 1161, in do_install
    rec = spack.store.db.get_record(self.spec)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/database.py", line 81, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/database.py", line 554, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/database.py", line 549, in _get_matching_spec_key
    raise KeyError("No such spec in database! %s" % spec)
KeyError: 'No such spec in database! mpich@3.2%intel@17.0.0 cflags="-axCORE-AVX2,AVX" cxxflags="-axCORE-AVX2,AVX" fflags="-axCORE-AVX2,AVX"+hydra+pmi+romio+verbs arch=linux-centos6-x86_64'
```
Checked and I actually do have an `mpich@3.2%intel-17.0.0` installation directory with only a `.spack` directory inside. Had to delete it and run `spack reindex` to get back to normal. Seems like Spack isn't cleaning up properly after failed/killed installs.

**Comment by alalazo** [comment]
@adamjstewart @tgamblin I may have a suspect for this behavior. We have over the code-base a few try/except blocks that goes like:
```python
try:
   ....
except:
   ...
```
instead of at least
```python
try:
   ....
except Exception:
   ...
```
The bare `except:` statement catches `BaseException` instances which include `KeyboardInterrupt` exceptions. If somebody hits <kbd>Ctrl</kbd>+<kbd>C</kbd> when in the `try` part of the above blocks the clean-up procedure won't be triggered and hence the problem above.

**Comment by adamjstewart** [comment]
This happened to me again today. I hit <kbd>Ctrl</kbd>+<kbd>C</kbd> while installing `r-httpuv` and saw the following output:
```
$ spack install r-httpuv
...
Ctrl+C
...
Process logger_daemon:
Traceback (most recent call last):
  File "/usr/lib64/python2.6/multiprocessing/process.py", line 232, in _bootstrap
    self.run()
  File "/usr/lib64/python2.6/multiprocessing/process.py", line 88, in run
    self._target(*self._args, **self._kwargs)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/llnl/util/tty/log.py", line 177, in _spawn_writing_daemon
==> Error: Could not remove prefix /blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/gcc-6.1.0/r-httpuv-1.3.3-uopa2m6v2e63tl26o2jxkn3ejuwlguxj for r-httpuv@1.3.3%gcc@6.1.0 arch=linux-centos6-x86_64 -uopa2m6 : [Errno 2] No such file or directory: '/blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/gcc-6.1.0/r-httpuv-1.3.3-uopa2m6v2e63tl26o2jxkn3ejuwlguxj/rlib/R/library/httpuv/Meta'
    [read_file, input_stream], [], [], 0
KeyboardInterrupt
$ spack install r-httpuv
==> Error: Install prefix exists but contains no spec.yaml:
  /blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/gcc-6.1.0/r-httpuv-1.3.3-uopa2m6v2e63tl26o2jxkn3ejuwlguxj
```

**Comment by alalazo** [comment]
@adamjstewart Spack was informing you that it could not remove the prefix:
```
==> Error: Could not remove prefix /blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/gcc-6.1.0/r-httpuv-1.3.3-uopa2m6v2e63tl26o2jxkn3ejuwlguxj for r-httpuv@1.3.3%gcc@6.1.0 arch=linux-centos6-x86_64 -uopa2m6 : [Errno 2] No such file or directory: '/blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/gcc-6.1.0/r-httpuv-1.3.3-uopa2m6v2e63tl26o2jxkn3ejuwlguxj/rlib/R/library/httpuv/Meta'
    [read_file, input_stream], [], [], 0
```
I think the only thing you can do is remove it manually. 

Too bad we don't have the complete stack of exceptions (which comes with python 3), but I suspect that for some reason the prefix removal failed and there was an exception thrown when handling the `KeyboardInterrupt` exception.

**Comment by adamjstewart** [comment]
I can actually reliably reproduce this one! Right after the build crashes, it hangs for a few seconds, probably trying to delete the installation directory. If I kill it during this time, it causes Spack to think it is still installed. If I don't kill it, or kill it before then, Spack cleans up properly. I think we are not handling Ctrl+C properly when cleaning up.

**Comment by alalazo** [comment]
> I think we are not handling Ctrl+C properly when cleaning up.

I am afraid we don't handle at all `KeyboardInterrupt` s when they come from an exceptional path...

**Comment by adamjstewart** [comment]
Reproduced this again:
```
...
==> Successfully installed util-macros
  Fetch: 0.48s.  Build: 1.37s.  Total: 1.85s.
[+] /blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/pgi-16.5-0/util-macros-1.19.1-jridl43rw4rhbnghcqdqqo72xhdavpbn
^C
==> Error: Keyboard interrupt.
$ spack install libpciaccess
==> Installing libpciaccess
==> util-macros is already installed in /blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/pgi-16.5-0/util-macros-1.19.1-jridl43rw4rhbnghcqdqqo72xhdavpbn
Traceback (most recent call last):
  File "/soft/spack-0.10.0/bin/spack", line 212, in <module>
    main(sys.argv)
  File "/soft/spack-0.10.0/bin/spack", line 208, in main
    _main(args, unknown)
  File "/soft/spack-0.10.0/bin/spack", line 174, in _main
    return_val = command(parser, args)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/cmd/install.py", line 346, in install
    package.do_install(**kwargs)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/package.py", line 1190, in do_install
    **kwargs
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/package.py", line 1162, in do_install
    rec = spack.store.db.get_record(self.spec)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/database.py", line 81, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/database.py", line 554, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/database.py", line 549, in _get_matching_spec_key
    raise KeyError("No such spec in database! %s" % spec)
KeyError: 'No such spec in database! util-macros@1.19.1%pgi@16.5-0 cflags="-tp=haswell,sandybridge" cxxflags="-tp=haswell,sandybridge" fflags="-tp=haswell,sandybridge"  arch=linux-centos6-x86_64'
```
You'll notice that this time I (unintentionally) killed Spack immediately after it printed out the message saying it was installed. Spack must have quit without adding it to the database.

**Comment by alalazo** [comment]
I think this is a different problem than before. Before, you were probably giving <kbd>Ctrl</kbd>+<kbd>C</kbd> while here:
```python
if os.path.exists(path):
    try:
        shutil.rmtree(path) # <--
    except exceptions.OSError as e:
        raise RemoveFailedError(spec, path, e)
```
which is called as part of an exceptional path:
```python
        try:
            # Create the install prefix and fork the build process.
            spack.store.layout.create_install_directory(self.spec)
            # Fork a child to do the actual installation
            spack.build_environment.fork(self, build_process, dirty=dirty)
            # If we installed then we should keep the prefix
            keep_prefix = True if self.last_phase is None else keep_prefix
            # note: PARENT of the build process adds the new package to
            # the database, so that we don't need to re-read from file.
            spack.store.db.add(
                self.spec, spack.store.layout, explicit=explicit
            )
        except directory_layout.InstallDirectoryAlreadyExistsError:
            # Abort install if install directory exists.
            # But do NOT remove it (you'd be overwriting someone else's stuff)
            tty.warn("Keeping existing install prefix in place.")
            raise
        except StopIteration as e:
            # A StopIteration exception means that do_install
            # was asked to stop early from clients
            tty.msg(e.message)
            tty.msg(
                'Package stage directory : {0}'.format(self.stage.source_path)
            )
        finally:
            # Remove the install prefix if anything went wrong during install.
            if not keep_prefix:
                self.remove_prefix()  # <---
```
and this is not handled gracefully. I still have to figure out what could be the sequence causing the last problem you posted.

**Comment by adamjstewart** [comment]
Yeah, I think these are at least 3 different problems, all with similar symptoms as a result. I know most of these are crazy, but I don't want other users to run into these and not know how to solve them.

**Comment by alalazo** [comment]
>  I know most of these are crazy, but I don't want other users to run into these and not know how to solve them.

I agree. Now that the basic features of Spack are more solid we need to focus a bit more on polishing the UI, which among other things means not having behaviors like these. 

And never fail with a stacktrace, just reassuring messages  :laughing:

**Comment by jrood-nrel** [comment]
So I've been experiencing this more lately and people I have using Spack are seeing it as well. It's actually most troublesome for myself now since I have a test framework running nightly and when Spack thinks something is already installed, it breaks everything and I have to intervene by merely deleting the install prefix. I don't know the code, but is it possible to delay the creation of the install prefix until the last step? It appears to be created at build right now?

**Comment by alalazo** [comment]
@jrood-nrel Which kind of failure do you have at install time that is not handled by the `finally` clause in the code posted above?

**Comment by jrood-nrel** [comment]
To be honest I don't really know. It's difficult to keep track of operations, but when a build just fails for whatever reason, sometimes I have to go remove an install prefix myself. I guess I feel like Trilinos is especially susceptible to this since it takes so long to build, the empty prefix is just sitting there waiting to give me an error on my next build attempt if it fails.

**Comment by adamjstewart** [comment]
I've ran into this bug several times this week. It seems like every time I try to install something, hit <kbd>Ctrl</kbd>+<kbd>C</kbd> during configure, and re-run `spack install`, it crashes.

Normally (in the past) when I hit <kbd>Ctrl</kbd>+<kbd>C</kbd>, Spack would print an error message starting with `==>` mentioning that it caught my signal. Now it simply prints `Terminated`.

**Comment by jrood-nrel** [comment]
Almost all of my builds are as batch jobs and I encounter this several times a week. It's currently my least favorite thing about Spack.

**Comment by adamjstewart** [comment]
Just did it again. I think I can reliably reproduce this now if anyone has any fixes they want tested.

**Comment by scheibelp** [comment]
I'm self-assigning this. I can start looking at the issue this week (and hopefully a PR by this Friday). If someone else was looking at this let me know.

I'm thinking the most straightforward approach would be to create an additional file in the ```.spack``` metadata directory in the package prefix to indicate that the install (and post-install hooks) completed (e.g. ```.spack/install_marker```, which could be empty). If a command line option is enabled then a package prefix will be removed whenever it does not contain this file.

I think it would also be possible to alter the SIGINT handler during the installation (starting before the point the prefix is created) but it would be harder to do so in a straightforward manner.

@tgamblin @becker33 curious if you have thoughts/preference between the above two strategies.

**Comment by tgamblin** [comment]
@scheibelp: great!

You're always going to have cases where something can `kill -9` a build and force it to die without cleanup.  So the case needs to be handled regardless of whether we improve SIGINT handling or not.

Touching an install marker in `.spack` (maybe `.spack/complete`?) seems like a reasonable way to go.  Personally, I think that we should remove incomplete install prefixes on reinstall *by default*, if not unconditionally.  Also, it would be good to establish one of `DirectoryLayout.check_installed()` or `Package.installed` as the source of truth on this.  I think this logic (and the associated file convention) should probably be managed by the directory layout, and `Package.installed` should call something in `DirectoryLayout`.

My first stab at having `check_installed` do detailed checks and `Package.installed` do a simple existence test for the prefix has probably caused most of these errors.

**Comment by adamjstewart** [comment]
> Personally, I think that we should remove incomplete install prefixes on reinstall by default, if not unconditionally.

The only time I would prefer not to do this is if I run `spack install --run-tests`, the tests fail, but I don't really care and just want to install it. In that case, for most packages, it will quickly finish up the last steps. But for some packages it will cause a problem if you run configure without cleaning the directory first, so I can see why we would want to default to cleaning up.

**Comment by tgamblin** [comment]
@adamjstewart>:
The only time I would prefer not to do this is if I run spack install --run-tests, the tests fail, but I don't really care and just want to install it.

There's a more complicated discussion here on tests that run in the build directory, tests that run from an install directory, etc. and how we should record whether tests pass.  I'd like to factor `--run-tests` out into some separate methods for different packages and rework the `spack test` command to be able to test packages.  Then we could have separate ``.spack/status/installed`, `.spack/status/tested`, etc. files.  It wouldn't be horrible to store that stuff in the DB either (though i like to keep a backup in the `.spack` directories in case we need to regenerate the DB).

Anyway, it sounds like you want unconditional removal by default, with an option to keep.  Would it make sense to reuse the `--keep-prefix` option for this?  Currently it keeps the prefix on install failure.  This would just make it also *keep* keeping the prefix after a prior build's failure.

**Comment by adamjstewart** [comment]
> Anyway, it sounds like you want unconditional removal by default, with an option to keep.

This is just a use case. I really wouldn't be offended if we cleaned up by default and there was no option to preserve the files. But `--keep-prefix` sounds good to me as far as UI goes.

**Comment by tgamblin** [comment]
Ok @scheibelp: let's use `--keep-prefix` to skip the removal if needed.

**Comment by tgamblin** [comment]
Reopening pending fix for issues in #3530.
